Match Roslyn EE to native EE behaviour with property accessibility

saul opened the pull request:
**Customer scenario**

The Roslyn EE breaks debug proxy types in FSharp.Core, meaning that maps, lists and sets are near impossible to debug as the Roslyn EE thinks all of the properties on the debug proxy types are not public. Roslyn EE is correct - the properties aren't public - but the Native EE accepted this. It's important to note that properties on internal types in F# are *impossible* to make public.

I believe that as the Roslyn EE causes a regression with F# compared to the Native EE, the old behaviour should be used instead.

See https://github.com/Microsoft/visualfsharp/issues/2184#issuecomment-292645097. 

**Before**:

![image](https://cloud.githubusercontent.com/assets/64654/24877897/d3a7eb28-1e28-11e7-8dfe-d9f0b05f49e8.png)

**After**:

![image](https://cloud.githubusercontent.com/assets/64654/24877842/a907e6c0-1e28-11e7-8d96-a34610e6b210.png)

**Bugs this fixes:**

Fixes #18581, fixes https://github.com/Microsoft/visualfsharp/issues/2184

**Workarounds, if any**

**Risk**

**Performance impact**

 "Low"

**Is this a regression from a previous update?**

**Root cause analysis:**

**How did we miss it? What tests are we adding to guard against it in the future?**

**How was the bug found?**

customer reported

tmat commented:
👎

This change skips the visibility check completely. With the change even entirely private property in C# will be shown. I'd be more willing to accept a change that displays the property if at least one of the acessors are visible. Even then it feels like a hack -- we should not compensate for bugs in F# compiler in C# EEs. 

I'd rather see F# having its own EE that implements the workaround. Roslyn provides the base class for Result Provider from which C# and VB providers inherit. Step one would be to inherit F# result provider -- it can copy the C# one for now and implement F# semantics later on.

dsyme commented:
> This change skips the visibility check completely. 

Why is the visibility check even there?  Why would you want to stop a valid debug proxy working?    

Also, is there a public spec that says you must have the "public" on the method of an otherwise internal type?

dsyme commented:
@tmat Also, would an F# EE actually be used when in a C# code file, looking at these values (in C#-calls-F# scenarios)?

saul commented:
I'd just like to add that I think Roslyn EE is doing the right thing here. It's how I would write it - it makes sense.

However I think that as this PR means that private getter, public setter properties are shown on types in the debugger (how common are these?), restoring an old behaviour and removing no functionality, is significantly outweighed by the fact that F# users cannot view the contents of collections.

tmat commented:
@dsyme @saul I gathered this is a bug in F# compiler from the discussion on https://github.com/Microsoft/visualfsharp/issues/2184#issuecomment-292645097:

> I think this is something we need to fix from the F# compiler side. I'm just awaiting @dsyme's comment, as I think we're setting the accessibility of property getter/setter methods incorrectly.

But you make good points about debugging older F# bits and looking at F# objects from C# method. As I said, I'd be willing to accept the change that displays the property if at least one of the accessors is visible. That seems to be what native EE actually does based on my observation:

```
    [DebuggerTypeProxy(typeof(X))]
    public class C { }

    internal class X
    {
        internal X(C c) { }
        private int Private { get; set; }
        internal int Internal { get; set; }
        public int Public { get; set; }
        internal int InternalPrivate { get; private set; }
        internal int PrivateInternal { private get; set; }
        public int PrivatePublic { private get; set; }
        public int PublicPrivate { get; private set; }
    }
```

Roslyn EE:
![image](https://cloud.githubusercontent.com/assets/41759/25014005/d998d5fc-202a-11e7-8c9b-11efa5301887.png)

Legacy EE:
![image](https://cloud.githubusercontent.com/assets/41759/25014013/ddc99fb2-202a-11e7-8c3b-b11262b020f1.png)

saul commented:
@tmat I don't think your conclusion is entirely true re how the native EE handles this.

Looking at the IL for Microsoft.FSharp.Collections.MapDebugView`2<TKey, TValue>, it doesn't have any public getter/setters for the Items property, yet it appears fine in the native EE debugger view.

```
	.method assembly hidebysig specialname 
		instance valuetype [mscorlib]System.Collections.Generic.KeyValuePair`2<!TKey, !TValue>[] get_Items () cil managed

	.property instance valuetype [mscorlib]System.Collections.Generic.KeyValuePair`2<!TKey, !TValue>[] Items()
	{
		.custom instance void [mscorlib]System.Diagnostics.DebuggerBrowsableAttribute::.ctor(valuetype [mscorlib]System.Diagnostics.DebuggerBrowsableState) = (
			01 00 03 00 00 00 00 00
		)
		.get instance valuetype [mscorlib]System.Collections.Generic.KeyValuePair`2<!0, !1>[] Microsoft.FSharp.Collections.MapDebugView`2::get_Items()
	}

```

saul commented:
It looks like having a `DebuggerBrowsable` attribute on the property makes the native EE ignore any accessibility constraints, try this out:

```c#
    struct Something
    {
        public int Foo;
    }

    [DebuggerTypeProxy(typeof(X))]
    public class C { }

    internal class X
    {
        internal X(C c) { }

        [DebuggerBrowsable(DebuggerBrowsableState.RootHidden)]
        internal Something Internal { get { return new Something(); } }
    }
```

tmat commented:
@saul You are right. Presence of the attribute overrides the visibility check and the native EE shows the return value of the property getter for all properties that have a getter regardless of their visibility. That's not however what your change does. The change needs to be done in the caller of the IsPublic member where it should check for the attribute. If the attribute is not present the current IsPublic method should be used as is. Otherwise ```m => true``` should be passed in as a member filter.

saul commented:
@tmat Agreed - I'll get working on that now.

saul commented:
Ready for review - any pointers as to where I should place some tests for this?

tmat commented:
test this please

saul commented:
@tmat I've fixed the existing tests and added a new test to verify the behaviour is as we expect. Expecting this to go green.

It's worth noting that I found that all of the DebuggerBrowsableAttributeTests tests are skipped at the moment because the class is internal. Is this on purpose?

tmat commented:
> It's worth noting that I found that all of the DebuggerBrowsableAttributeTests tests are skipped at the moment because the class is internal. 

@saul Good catch. Not intentional, could you please make them public?

tmat commented:
/cc @cston 
👍

tmat commented:
test windows_debug_vs-integration_prtest please
test windows_release_vs-integration_prtest please
https://github.com/dotnet/roslyn/issues/18714

cston commented on src/ExpressionEvaluator/CSharp/Test/ResultProvider/DebuggerTypeProxyAttributeTests.cs:
```diff
@@ -1100,5 +1100,50 @@ public P(C c)
                 EvalResult("F", "3", "int", "(new C(3)).F"));
         }
+
+        [Fact]
```
Please add `[WorkItem(18581, "https://github.com/dotnet/roslyn/issues/18581")]`.

cston commented:
/cc @jinujoseph, @ivanbasov

cston commented:
LGTM. Thanks for fixing this @saul.

saul commented:
@tmat I'll make the tests public again but one is failing - shall I change it so it passes?

Also I'm not sure why the VB symbol highlighting integration test is now failing - I've no idea where to even start with that. Any pointers would be appeciated as it doesn't look flaky.

tmat commented:
Yes, please fix the failing test, unless the results don't seem correct.

You can ignore the failing integration test.

saul commented:
@tmat I've done the review comments and fixed the one broken DebuggerBrowsableAttributeTest.

cston commented on src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/MemberExpansion.cs:
```diff
@@ -26,7 +26,8 @@ internal sealed class MemberExpansion : Expansion
             Predicate<MemberInfo> predicate,
-            ResultProvider resultProvider)
+            ResultProvider resultProvider,
+            bool isProxyType)
```
Did an existing test fail without `isProxyType`? If not, please add a test that demonstrates the difference.

saul replied:
Committed a test to ensure that the behaviour on non-proxy types is not affected by this PR in 97164c9.

cston commented:
LGTM

jinujoseph commented:
test windows_debug_vs-integration_prtest please

jinujoseph commented:
@MattGertz fyi

tmat commented:
@MattGertz Clicked merge before realizing ask mode approval is needed.

dsyme commented:
Thank you everyone for helping to resolve this issue, really great to see the cooperation and eventual precision in the fix
